chain/chain/src/stateless_validation/chunk_validation.rs

```diff
@@ -549,43 +637,67 @@ pub fn validate_chunk_state_witness(
         );
     }
 
-    for (block, transition) in pre_validation_output
+    if pre_validation_output.implicit_transition_params.len()
+        != state_witness.implicit_transitions.len()
+    {
+        return Err(Error::InvalidChunkStateWitness(format!(
+            "Implicit transitions count mismatch. Expected {}, found {}",
+            pre_validation_output.implicit_transition_params.len(),
+            state_witness.implicit_transitions.len(),
+        )));
+    }
+
+    for (implicit_transition_params, transition) in pre_validation_output
```

Comment:
Just for my education - what exactly is the length of the params, the pre validation and why are they always equal?

Reply:
ChunkStateWitness is expected to provide all implicit state transitions, but it must be validated.
Chunk validator can retrieve number of implicit transitions by chain traversal itself. It is always number of missing chunks plus resharding transition if needed. If the loop completes, it means that implicit transitions were provided correctly.